airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGeneratorIntegrationTest.java

```diff
@@ -369,4 +372,24 @@ record -> record.get(JavaBaseConstants.COLUMN_NAME_AB_RAW_ID).asText(),
     });
   }
 
+  /**
+   * We test this for Snowflake because we made a special exception class, _ab_missing_primary_key
+   */
+  @Test
+  protected void customExceptionWhenMissingPK() throws Exception {
+    createRawTable(streamId);
+    createFinalTable(incrementalDedupStream, "");
+    insertRawTableRecords(
+        streamId,
+        BaseTypingDedupingTest.readRecords("sqlgenerator/incrementaldedup_inputrecords_missing_pk.jsonl"));
+
+    final String sql = generator.updateTable(incrementalDedupStream, "");
+    Exception exception = assertThrows(Exception.class, () -> { // TODO (evan): I don't really know what class of error the Snowflake driver throws
+      destinationHandler.execute(sql);
+    });
+
+    assertTrue(exception.getMessage().contains("_AB_MISSING_PRIMARY_KEY"));
+    assertTrue(exception.getMessage().contains("\"users_raw\"` has rows missing a primary key"));
```

Comment:
this is sort of redundant with https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/bases/base-typing-deduping-test/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/BaseSqlGeneratorIntegrationTest.java#L334

(your new test adds asserts on the specific error message though, which is probably valuable? maybe makes sense to actually override the existing test case)